src/ServiceControl.Transports.RabbitMQ/QueueLengthProvider.cs

```diff
@@ -0,0 +1,177 @@
+
+            return queryLoop;
+        }
+
+        void UpdateQueueLengths()
+        {
+            var nowTicks = DateTime.UtcNow.Ticks;
+
+            foreach (var endpointQueuePair in endpointQueues)
+                store.Store(
+                    new[]
+                    {
+                        new RawMessage.Entry
+                        {
+                            DateTicks = nowTicks,
+                            Value = sizes.TryGetValue(endpointQueuePair.Value, out var size) ? size : 0
+                        }
+                    },
+                    endpointQueuePair.Key);
+        }
+
+        async Task FetchQueueLengths()
+        {
+            foreach (var endpointQueuePair in endpointQueues)
+            {
+                await queryExecutor.Execute(m =>
+                {
+                    var queueName = endpointQueuePair.Value;
+
+                    try
+                    {
+                        var size = (int) m.MessageCount(queueName);
+
+                        sizes.AddOrUpdate(queueName, _ => size, (_, __) => size);
+                    }
+                    catch (Exception e)
+                    {
+                        Logger.Warn($"Error quering queue length for {queueName}", e);
+                    }
+                });
+            }
+        }
+
+
+        class QueryExecutor
+        {
+            string connectionString;
+            IConnection connection;
+            IModel model;
+            ConnectionFactory connectionFactory;
+
+            public QueryExecutor(string connectionString)
+            {
+                this.connectionString = connectionString;
+            }
+
+            public void Initialize()
+            {
+                var connectionConfiguration = ConnectionConfiguration.Create(connectionString, "ServiceControl.Monitoring");
+                connectionFactory = new ConnectionFactory(connectionConfiguration, null, false, false);
+            }
+
+            public async Task Execute(Action<IModel> action)
+            {
+                try
+                {
+                    if (connection == null)
+                    {
+                        connection = connectionFactory.CreateConnection(string.Empty);
+                    }
+
+                    //Connection implements reconnection logic
+                    while (connection.IsOpen == false)
+                    {
+                        await Task.Delay(ReconnectionDelay);
+                    }
+
+                    if (model == null || model.IsClosed)
+                    {
+                        model = connection.CreateModel();
+                    }
+
+                    action(model);
+                }
+                catch (Exception e)
+                {
+                    Logger.Warn("Error querying queue length.", e);
+                }
+            }
+
+            TimeSpan ReconnectionDelay = TimeSpan.FromSeconds(5);
```

Comment:
This is _forever_. What if the reconnection happens just after we checked the flag?

Reply:
This is based on the fact that 5 sec is the default [timeout](https://www.rabbitmq.com/dotnet-api-guide.html#connection-recovery) for the built-in rabbitmq client library recoverability.

Reply:
Actually the default used by RabbitMQ client is 10 sec. See @bording  [comment](https://github.com/Particular/ServiceControl.Monitoring/pull/133#discussion_r193945454)

Reply:
10 seconds is the default that the transport uses when creating a `ConnectionFactory`, but the client's default value if the `NetworkRecoveryInterval` property isn't modified is 5 seconds.